# Unpacking Instacart: A Deep Dive into North American Grocery E-Commerce Behavior

## Preliminary Data Exploration Analysis

The **main aim** is to get a basic feel of the data and understand the relationships inside a dataset before its uploading to SQL.

**Tasks:**
* Check for missing values and duplicates.
* Check dimensions of the dataset.
* Understand the type of variables present.

### Import Statements

In [2]:
import pandas as pd

### Read the CSV files into a DataFrame

In [18]:
df_aisles = pd.read_csv('data/aisles.csv')
df_departments = pd.read_csv('data/departments.csv')
df_orders_prior = pd.read_csv('data/order_products__prior.csv')
df_orders_train = pd.read_csv('data/order_products__train.csv')
df_orders = pd.read_csv('data/orders.csv')
df_products = pd.read_csv('data/products.csv')
df_sample_submission = pd.read_csv('data/sample_submission.csv')

### Check for Missing Values and Duplicates

In [19]:
dataframes = [df_aisles, df_departments, df_products, df_orders, df_orders_prior, df_orders_train, df_sample_submission]

for df in dataframes:
    df_name = [name for name, obj in locals().items() if obj is df][0]
    print(f'{df_name} has duplicates: {df.duplicated().values.any()}')
    print(f'{df_name} has NaN values: {df.isna().values.any()}')

df_aisles has duplicates: False
df_aisles has NaN values: False
df_departments has duplicates: False
df_departments has NaN values: False
df_products has duplicates: False
df_products has NaN values: False
df_orders has duplicates: False
df_orders has NaN values: True
df_orders_prior has duplicates: False
df_orders_prior has NaN values: False
df_orders_train has duplicates: False
df_orders_train has NaN values: False
df_sample_submission has duplicates: False
df_sample_submission has NaN values: False


* There are no duplicates in all dataframes.
* However, `df_orders` dataframe has missing values values. This needs to be investigated further.

In [20]:
# Identifying column(s) with missing values
df_orders.isna().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [27]:
df_orders.groupby('eval_set').count()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
eval_set                                                                  
prior      3214874  3214874       3214874    3214874            3214874   
test         75000    75000         75000      75000              75000   
train       131209   131209        131209     131209             131209   

          days_since_prior_order  
eval_set                          
prior                    3008665  
test                       75000  
train                     131209

* The only column with 206,209 NaN values is `'days_since_prior_order'` in the `'prior'` evaluation set.
* NaN values in this column mean the first order by a user and are not connected with any issue.
* **75,000 orders** in `'test'` data are reserved for machine learning competition. No product data in this evaluation set is expected.

### Check Dimensions of the Dataset and Data Type of Variables

### DataFrame #1: Products

In [32]:
print(df_products.info())
df_products.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49688 entries, 0 to 49687
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   product_id     49688 non-null  int64 
 1   product_name   49688 non-null  object
 2   aisle_id       49688 non-null  int64 
 3   department_id  49688 non-null  int64 
dtypes: int64(3), object(1)
memory usage: 1.5+ MB
None


product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  
0             19  
1             13  
2              7  
3              1  
4             13

* No issues with the data type.
* There are **49,688 products** available on the platform.

### DataFrame #2: Aisles

In [28]:
print(df_aisles.info())
df_aisles.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 134 entries, 0 to 133
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   aisle_id  134 non-null    int64 
 1   aisle     134 non-null    object
dtypes: int64(1), object(1)
memory usage: 2.2+ KB
None


aisle_id                       aisle
0         1       prepared soups salads
1         2           specialty cheeses
2         3         energy granola bars
3         4               instant foods
4         5  marinades meat preparation

* No issues with the data type.
* Products on the platform are organized into **134 aisles**.

### DataFrame #3: Departments

In [31]:
print(df_departments.info())
df_departments.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21 entries, 0 to 20
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   department_id  21 non-null     int64 
 1   department     21 non-null     object
dtypes: int64(1), object(1)
memory usage: 464.0+ bytes
None


department_id department
0              1     frozen
1              2      other
2              3     bakery
3              4    produce
4              5    alcohol

* No issues with the data type.
* Aisles on the platform are grouped into **21 departments**.

### DataFrame #4: Orders

In [33]:
print(df_orders.info())
df_orders.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(5), object(1)
memory usage: 182.7+ MB
None


order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [26]:
# User count and their orders count
df_orders.groupby('user_id').count().sort_values('order_id', ascending=False)

order_id  eval_set  order_number  order_dow  order_hour_of_day  \
user_id                                                                   
152340        100       100           100        100                100   
185641        100       100           100        100                100   
185524        100       100           100        100                100   
81678         100       100           100        100                100   
70922         100       100           100        100                100   
...           ...       ...           ...        ...                ...   
13046           4         4             4          4                  4   
70208           4         4             4          4                  4   
111440          4         4             4          4                  4   
31762           4         4             4          4                  4   
47980           4         4             4          4                  4   

         days_since_prior_order  
user_id                          
152340                       99  
185641                       99  
185524                       99  
81678                        99  
70922                        99  
...                         ...  
13046                         3  
70208                         3  
111440                        3  
31762                         3  
47980                         3  

[206209 rows x 6 columns]

In [29]:
# Unique values in `order_dow` column
df_orders.order_dow.unique()

array([2, 3, 4, 1, 5, 0, 6])

* No issues with the data type.
* This file tells to which set (prior, train, test) an order belongs (based on `'eval_set'` column) and is supposed to be a central file in the dataset.
* **3,421,083 orders** were placed on the platform.
* The dataset is anonymized (`'user_id'` only) and contains grocery orders from **206,209 Instacart users**.
    * Previously, in the `'days_since_prior_order'` column of the 'prior' evaluation set 206,209 NaN values were identified. Notably, this number matches the count of unique users. This supports the idea that these missing values represent first-time orders and there is no issue with the data.
* For each user, **between 4 and 100 of their orders** are provided.
* `'order_dow'` is the day of week the order was placed on. It is represented by a number from 0 to 6. A week in the US starts from Sunday, that's why I have a hypothesis that 0 stands for Sunday, 1 for Monday, etc.
* `'order_number'`: the order sequence number for this user (1 = first, n = nth).
* There is no information about products sold in each order.

### DataFrame #5: Orders Prior

In [35]:
print(df_order_products_prior.info())
df_order_products_prior.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32434489 entries, 0 to 32434488
Data columns (total 4 columns):
 #   Column             Dtype
---  ------             -----
 0   order_id           int64
 1   product_id         int64
 2   add_to_cart_order  int64
 3   reordered          int64
dtypes: int64(4)
memory usage: 989.8 MB
None


order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

* No issues with the data type.
* This file contains **previous order** contents for all customers and **specifies which products** were purchased in each order (there is a reference to `product_id`).
* `'reordered'` indicates that the customer has a previous order that contains the product:
    * 1 if this product has been ordered by this user in the past,
    * 0 otherwise.
* Memory usage is nearly 1GB; consider this when merging with other data frames.

### DataFrame #6: Orders Train

In [36]:
print(df_order_products_train.info())
df_order_products_train.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1384617 entries, 0 to 1384616
Data columns (total 4 columns):
 #   Column             Non-Null Count    Dtype
---  ------             --------------    -----
 0   order_id           1384617 non-null  int64
 1   product_id         1384617 non-null  int64
 2   add_to_cart_order  1384617 non-null  int64
 3   reordered          1384617 non-null  int64
dtypes: int64(4)
memory usage: 42.3 MB
None


order_id  product_id  add_to_cart_order  reordered
0         1       49302                  1          1
1         1       11109                  2          1
2         1       10246                  3          0
3         1       49683                  4          0
4         1       43633                  5          1

* No issues with the data type.
* This file identifies **the last order for 131,209 users** (= 206,209 - 75,000) and **specifies which products** were purchased in each order (there is a reference to `product_id`).
* `'reordered'` indicates that the customer has a previous order that contains the product:
    * 1 if this product has been ordered by this user in the past,
    * 0 otherwise.

### DataFrame #7: Orders Test (Sample Submission)

In [34]:
print(df_sample_submission.info())
df_sample_submission.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 75000 entries, 0 to 74999
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   order_id  75000 non-null  int64 
 1   products  75000 non-null  object
dtypes: int64(1), object(1)
memory usage: 1.1+ MB
None


order_id     products
0        17  39276 29259
1        34  39276 29259
2       137  39276 29259
3       182  39276 29259
4       257  39276 29259

* No issues with the data type.
* This file identifies **the last order for 75,000 users** and represents the 'test' set for machine learning competition.
* `'products'` column contains an example of a space-delimited list of product_ids for that order. Thus, for further data analysis purporses this file will be out of scope.

In [25]:
# Product hierarchy
print(f'The dataset has {df_products.shape[0]} products, which are grouped in {df_aisles.shape[0]} aisles and {df_departments.shape[0]} departments.')

# Number of orders
print(f'The dataset contains {df_orders.shape[0]} grocery orders.')

The dataset has 49688 products, which are grouped in 134 aisles and 21 departments.
The dataset contains 3421083 grocery orders.


## Findings & Conclusions

#### Dataset structure
* The dataset is a **relational set of 7 files** describing customers' orders over time.
* Three of them (`departments.csv`, `aisles.csv`, `products.csv`) contain product related information only.
* The other four contain information about users' orders:
    * `orders.csv` is the central file with an information about all orders made on the platform, but without product specifics.
    * `order_products__prior.csv` and `order_products__prior.csv` specifies which products were purchased in each order.
    * `sample_submission.csv` contains test data (order IDs) reserved for machine learning competitions. This file identifies **the last order for 75,000 users**. For further data analysis purporses this file will be **out of scope**.

#### Data cleaning is not needed:
* no data type issues were identified,
* no duplicated values,
* no unexpected missing values,
* data in all files looks logically neat.

#### Dataset dimensions overview

* **Orders & users**:
    * The dataset is anonymized and contains a sample of over 3 million (3,421,083) grocery orders from more than 200,000 (206,209) Instacart users.
    * For each user, provided between 4 and 100 of their orders, with the sequence of products purchased in each order.
* **Product hierarchy**:
    * There are 49,688 products available on the platform.
    * They are organized into 134 aisles and 21 departments.

* **Other**: 
    * The day of week and hour of day the order was placed, and a relative measure of time between orders, are provided.
    * The day of week when the order was placed on is represented by a number from 0 to 6. There is a hypothesis that 0 stands for Sunday, 1 for Monday, etc. (based on the USA calendar tradition).
    * Order sequence number for each user is also provided.
